README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rogme

Robust Graphical Methods For Group Comparisons (v. 0.2.1)

The `rogme` R package provides graphical tools and robust statistical
methods to compare groups of continous and pseudo-continuous
observations. The goal is to illustrate and quantify how and by how much
groups differ. The current version of the package is limited to
comparing two groups (though multiple pairs of groups can be compared in
one go). Future developments will extend the tools to deal with multiple
groups, interactions and hierarchical designs.

**NEW**: a [hierarchical shift function](docs/hsf.md) to compare two
dependent conditions from one group is now available. It has parametric
and bootstrap versions. The approach is described in [Rousselet &
Wilcox, 2019](https://psyarxiv.com/3y54r/).

The package can be installed using these commands:

``` r
install.packages("devtools")
devtools::install_github("GRousselet/rogme")
```

The approach behind the package can be summarised in one figure:

![](docs/Figure1.png) **How two independent distributions differ.
Left**: standard but misleading bar graphs of mean values. **Right**:
detailed graphical methods. **A.** Stripcharts of marginal
distributions. Vertical lines mark the deciles, with a thicker line for
the median. **B.** Kernel density representation and rug plot of the
distribution of difference scores. Vertical lines mark the deciles, with
a thicker line for the median. **C.** Shift function. Group 1 - Group 2
is plotted along the y-axis for each decile (white disk), as a function
of Group 1 deciles. The vertical lines indicate 95% bootstrap confidence
intervals. The shift function can be sparser or denser by changing the
quantiles. **D.** Difference asymmetry function with 95% bootstrap
confidence intervals.

The approach is also described in these articles:

[A few simple steps to improve the description of group results in
neuroscience](http://onlinelibrary.wiley.com/doi/10.1111/ejn.13400/full)

[Beyond differences in means: robust graphical methods to compare two
groups in
neuroscience](https://www.biorxiv.org/content/early/2017/03/27/121079).
\[[Reproducibility package using
`rogme`](https://figshare.com/articles/Modern_graphical_methods_to_compare_two_groups_of_observations/4055970)\]

`rogme` uses `ggplot2` for graphical representations, and the main
statistical functions were developed by Rand Wilcox, as part of his
[`WRS`](https://dornsife.usc.edu/labs/rwilcox/software/) package.

The main tool in `rogme` is the [shift
function](https://garstats.wordpress.com/2016/07/12/shift-function/). A
shift function shows the difference between the quantiles of two groups
as a function of the quantiles of one group. For inferences, the
function returns an uncertainty interval for each quantile difference.
By default, the deciles are used. Currently, confidence intervals are
computed using one of two percentile bootstrap techniques. Highest
density intervals and [Bayesian
bootstrap](https://github.com/rasmusab/bayesboot) intervals will be
available eventually.

## Vignettes

  - [Quantify distribution differences using the shift
    function](docs/sf_patterns.md)
  - [Compare two independent groups](docs/ind_gps.md)
  - [Compare two dependent groups](docs/dep_gps.md)
  - [Quantify a single distribution](docs/one_gp.md)
  - [Statistical tests and measures of effect sizes](docs/stats.md)
  - [Hierarchical shift function](docs/hsf.md)

## Functions

All the main functions rely on the [Harrell-Davis quantile
estimator](https://garstats.wordpress.com/2016/06/09/the-harrell-davis-quantile-estimator/),
computed by the `hd()` function.

### Shift function

In the `WRS` package, the shift function can be calculated using:

  - `shifthd()` or `qcomhd()` for independent groups
  - `shiftdhd()` or `Dqcomhd()` for dependent groups

These functions can also produce non-ggplot figures.

In `rogme`, the shift function can be calculated using:

  - `shifthd()` or `shifthd_pbci()` for independent groups
  - `shiftdhd()` or `shiftdhd_pbci()` for dependent groups

Illustrations of the results is handled separately by `plot_sf()`.

You can see the shift function in action for instance in these
publications:

  - [blog
    post](https://garstats.wordpress.com/2016/07/12/shift-function/)

  - [review article in the European Journal of
    Neuroscience](http://onlinelibrary.wiley.com/doi/10.1111/ejn.13100/full)

  - [Sex Differences in the Adult Human Brain: Evidence from 5216 UK
    Biobank
    Participants](https://academic.oup.com/cercor/advance-article/doi/10.1093/cercor/bhy109/4996558)

  - [Visibility graphs for fMRI data: Multiplex temporal graphs and
    their modulations across resting-state
    networks](https://www.mitpressjournals.org/doi/full/10.1162/NETN_a_00012)

  - [Manipulation of contact network structure and the impact on
    foot-and-mouth disease
    transmission](https://www.sciencedirect.com/science/article/pii/S0167587717305627)

  - [No evidence that frontal eye field tDCS affects latency or accuracy
    of
    prosaccades](https://www.biorxiv.org/content/early/2018/06/20/351304)

  - [A Key Characteristic of Sex Differences in the Developing Brain:
    Greater Variability in Brain Structure of Boys than
    Girls](https://www.ncbi.nlm.nih.gov/pubmed/28981610)

### Difference asymmetry function

The difference asymmetry function is another powerful graphical and
inferential tool. In the `WRS` package it is calculated using:

  - `qwmwhd()` for independent groups
  - `difQpci()` for dependent groups

In `rogme`, these functions have been renamed:

  - `asymhd()` for independent groups
  - `asymdhd()` for dependent groups
  - `plot_diff_asym()` to plot the results

You can see the difference asymmetry function in action in this [blog
post](https://garstats.wordpress.com/2016/07/19/typical-differences/)
and in this [review
article](http://onlinelibrary.wiley.com/doi/10.1111/ejn.13100/full).

## Shift function demo \[1\]

Detailed illustration of the shift function using two distributions that
differ in spread. The observations are in arbitrary units (a.u.).

``` r
#> generate data
set.seed(21)
g1 <- rnorm(1000) + 6
g2 <- rnorm(1000) * 1.5 + 6

#> make tibble
df <- mkt2(g1, g2)
```

First, we generate 1D scatterplots for the two groups.

``` r
#> scatterplots alone
ps <- plot_scat2(data = df,
                 formula = obs ~ gr,
                 xlabel = "",
                 ylabel = "Scores (a.u.)",
                 alpha = 1,
                 shape = 21,
                 colour = "grey10",
                 fill = "grey90") #> scatterplots
ps <- ps + coord_flip()
ps
```

![](README-files/README-unnamed-chunk-5-1.png)<!-- -->

Second, we compute the shift function and then plot it.

``` r
#> compute shift function
sf <- shifthd(data = df, formula = obs ~ gr, nboot = 200)
#> sf <- shifthd_pbci(data = df, formula = obs ~ gr, nboot = 200, q = c(.1,.25,.5,.75,.9))

#> plot shift function
psf <- plot_sf(sf, plot_theme = 2)
#> Warning: Using alpha for a discrete variable is not advised.

#> Warning: Using alpha for a discrete variable is not advised.

#> add labels for deciles 1 & 9
psf <- add_sf_lab(psf, sf, 
                  y_lab_nudge = .1, 
                  text_size = 4)

#> change axis labels
psf[[1]] <- psf[[1]] +  labs(x = "Group 1 quantiles of scores (a.u.)",
                             y = "Group 1 - group 2 \nquantile differences (a.u.)")
psf[[1]]
```

![](README-files/README-unnamed-chunk-6-1.png)<!-- -->

Third, we make 1D scatterplots with deciles and colour coded
differences.

``` r
p <- plot_scat2(df,
                xlabel = "",
                ylabel = "Scores (a.u.)",
                alpha = .3,
                shape = 21,
                colour = "grey10",
                fill = "grey90") #> scatterplots
p <- plot_hd_links(p, sf[[1]],
                    q_size = 1,
                    md_size = 1.5,
                    add_rect = TRUE,
                    rect_alpha = 0.1,
                    rect_col = "grey50",
                    add_lab = TRUE,
                    text_size = 5) #> superimposed deciles + rectangle
p <- p + coord_flip() #> flip axes
p
```

![](README-files/README-unnamed-chunk-7-1.png)<!-- -->

Finally, we combine the three plots into one figure.

``` r
library(cowplot)
cowplot::plot_grid(ps, p, psf[[1]], labels=c("A", "B", "C"), ncol = 1, nrow = 3,
                   rel_heights = c(1, 1, 1), label_size = 20, hjust = -0.5, scale=.95)
```

![](README-files/README-unnamed-chunk-8-1.png)<!-- -->

**Panel A** illustrates two distributions, both n = 1000, that differ in
spread. The observations in the scatterplots were jittered based on
their local density, as implemented in `ggbeeswarm::geom_quasirandom`.

**Panel B** illustrates the same data from panel A. The dark vertical
lines mark the deciles of the distributions. The thicker vertical line
in each distribution is the median. Between distributions, the matching
deciles are joined by coloured lined. If the decile difference between
group 1 and group 2 is positive, the line is orange; if it is negative,
the line is purple. The values of the differences for deciles 1 and 9
are indicated in the superimposed labels.

**Panel C** focuses on the portion of the x-axis marked by the grey
shaded area at the bottom of panel B. It shows the deciles of group 1 on
the x-axis – the same values that are shown for group 1 in panel B. The
y-axis shows the differences between deciles: the difference is large
and positive for decile 1; it then progressively decreases to reach
almost zero for decile 5 (the median); it becomes progressively more
negative for higher deciles. Thus, for each decile the shift function
illustrates by how much one distribution needs to be shifted to match
another one. In our example, we illustrate by how much we need to shift
deciles from group 2 to match deciles from group 1.

More generally, a shift function shows quantile differences as a
function of quantiles in one group. It estimates how and by how much two
distributions differ. It is thus a powerful alternative to the
traditional t-test on means, which focuses on only one, non-robust,
quantity. Quantiles are robust, intuitive and informative.

## Shift function demo \[2\]

The shift function can also be computed for all pairs of groups in a
data frame in one call by using the argument `doall = TRUE`.

``` r
set.seed(21) # generate data
n <- 100 # sample size
library(tibble)
df <- tibble(gr = factor(c(rep("group1",n),rep("group2",n),rep("group3",n))),
             obs= c(rnorm(n), rnorm(n)+1, rnorm(n)*5)) # make tibble
out <- shifthd(df, doall = TRUE) # compute all comparisons
```

Plot all shift functions in one call of `plot_sf`.

``` r
plist <- plot_sf(out)
```

The plots can then be combined using the packages
[gridExtra](https://cran.r-project.org/web/packages/gridExtra/index.html)
or
[cowplot](https://cran.r-project.org/web/packages/cowplot/index.html).

``` r
library(gridExtra)
do.call("grid.arrange", c(plist, ncol=2))
```

![](README-files/README-unnamed-chunk-11-1.png)<!-- -->

To extract one object and for instance change a label:

``` r
p <- plist[[1]]
p + labs(y = "Difference")
```

![](README-files/README-unnamed-chunk-12-1.png)<!-- -->

New group plot with different y labels and titles:

``` r
for(sub in 1:length(plist)){
plist[[sub]] <- plist[[sub]] + labs(y = "Difference", title = names(out)[sub]) + theme(plot.title = element_text(size = 20, face = "bold.italic"))
}
do.call("grid.arrange", c(plist, ncol=2))
```

![](README-files/README-unnamed-chunk-13-1.png)<!-- -->

To understand what’s going on, here are the marginal distributions:

``` r
p <- plot_scat2(df,
                xlabel = "",
                ylabel = "Scores (a.u.)",
                alpha = .5,
                shape = 21,
                colour = "grey10",
                fill = "grey90") #> scatterplots
p + coord_flip() #> flip axes
```

![](README-files/README-unnamed-chunk-14-1.png)<!-- -->


NEWS.md
# ------------------------
# rogme 0.2.2 (NOT RELEASED YET)
- new `onsets2` dataset, containing ERP onsets from 2 sessions
- fix bug in `plot_sf` when requesting quantiles other than the deciles
- fix bugs due to depricated `ggplot2` syntax in `plot_hd_bars` and `plot_hd_links`
- add missing `factor()` call in `mkt1` 

# ------------------------
# rogme 0.2.1 (2019-02-25)

- new `plot_hsf_pb_dist` function to plot percentile bootstrap samples
- new `plot_hsf_pb` function to plot percentile bootstrap results
- new `hsf_pb` function implementing percentile bootstrap hierarchical shift function
- new `hsf` function implementing basic hierarchical shift function
- new `plot_hsf` function to plot results
- add `flp.RData` to demonstrate hsf
- add `trimci`, `trimpval` and `winvar` functions to stats.R
- add `datasets.R` to document datasets
- rename `paired_example.RData` to `pdata.RData`
- rename `ERP_onsets.RData` to `onsets.RData`
- update vignettes `dep_gps` and `one_gp`
- create `hsf` vignette

# ------------------------
# rogme 0.2.0 (2018-07-07)

## General improvements
- add input checks to all main functions
- add unit tests for all main functions
- update help
- update README
- add 5 vignettes based on European Journal of Neuroscience 2017 paper examples
- remove redundant functions; make some functions more flexible

## New input/output for shift function and difference asymmetry function
- input data frames don't need to be created by `mkt1` or `mkt2` anymore
- more than 2 groups are allowed: by default first two groups are processed; options to process pairs
- change output to list of data frames
- matching plot functions updated to accept a list of data frames and return a list of ggplot objects
- the new format will make it easier to expand functions or create new ones to handle more complex designs

## Add formula input to:
- `plot_scat2`
- `plot_scat2d`

## Add handling of lists of ggplot objects to:
- `plot_sf`
- `add_sf_lab`

## New functions:
- `plot_hd_links` to replace `plot_dec_links`
- `plot_hd_ci`  to replace `plot_dec_ci`
- `subset_formula_wide`

## Delete functions:
- `deciles`
- `q1469`
- `plot_dec_bars`
- `plot_dec_links`
- `plot_quartile_bars`
- `plot_pbsf`
- `plot_dec_ci`
- `plot_scat2_sina`
- `annotate_quartiles`
- `subset_data2`
- `plot_kde_rug_dec1`
- `plot_kde_rug_dec2`

# ------------------------
# rogme 0.1.1 (2018-11-06) 

## Bug fixes

Changes suggested by John J Wood:
- Fix inconsistency between `mkt1` and `mkt2` column names:
mkt1: “data” , “gr”
mkt2: “obs”, “gr”
Now mkt1 creates a data frame with columns “obs” , “gr”
- Add function `q1469` to package
- In `plot_diff_asym` function, state that Wilcox’s `qwmwhd` and `difQpci` functions have been replaced with `asymhd` and `asymdhd`.

Bug spotted by zack0832:
- In `plot_kde_rug_dec1` and `plot_kde_rug_dec2`, replace "data" column with "obs" column from the `data` tibble.




R/marginal_plot.R
#' Plot one-dimensional scatterplots for 2 groups
#'
#' \code{plot_scat2} produces scatterplots for 2 marginal distributions.
#' The scatterplots are jittered using \code{\link[ggbeeswarm]{geom_quasirandom}}.
#'
#' @param data A data frame in long format. One column is a factor describing the groups;
#'   another column contains the values/observations for each group. A properly formatted data
#'   frame can be created using \code{\link{mkt2}}. Missing values are not
#'   allowed.
#' @param formula A formula with format response variable ∼ predictor variable,
#'   where ~ (tilde) means "is modeled as a function of".
#' @param xlabel Option to set different name - default NULL to use data frame column names.
#' @param ylabel Option to set different name - default NULL to use data frame column names.
#' @param ... Input arguments for ggbeeswarm::geom_quasirandom
#'
#' @return A ggplot object.
#'
#' @examples
#' # generate data
#' set.seed(21)
#' g1 <- rnorm(1000) + 6
#' g2 <- rnorm(1000) * 1.5 + 6
#'
#' # make tibble
#' df <- mkt2(g1, g2)
#' # make scatterplots
#' ps <- plot_scat2(data = df,
#'   formula = obs ~ gr,
#'   xlabel = "",
#'   ylabel = "Scores (a.u.)",
#'   alpha = 1,
#'   shape = 21,
#'   colour = "grey10",
#'   fill = "grey90") # scatterplots
#' ps <- ps + coord_flip()
#' ps
#'
#' @export
plot_scat2 <- function(data = df,
                       formula = obs ~ gr,
                       xlabel = NULL,
                       ylabel = NULL,
                       ...){
  # subset formula
  subf <- subset_formula(data, formula)
  xplot = subf$param_col_name
  yplot = subf$obs_col_name
  p <- ggplot(data, aes_string(x = xplot, y = yplot, fill = xplot,
                               colour = xplot, shape = xplot))
  p <- p + ggbeeswarm::geom_quasirandom(...) +
    theme_bw() +
    # scale_colour_manual(values = symb_col) +
    # scale_fill_manual(values = symb_fil) +
    # scale_shape_manual(values = symb_shape) +
    theme(legend.position = "none") +
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.title.y = element_text(size=16,face="bold"),
          axis.text.x = element_text(size=14),
          axis.text.y = element_text(size=14))
  # override axis labels
  if (!is.null(xlabel)){
    p <- p + xlab(xlabel)
  }
  if (!is.null(ylabel)){
    p <- p + ylab(ylabel)
  }
  p
}

#' Plot quantiles and confidence intervals
#'
#' Using the output of \code{\link{quantiles_pbci}}, create a ggplot object
#' showing specified quantiles (default to deciles) and their 95% percentile
#' bootstrap confidence intervals. A vertical line marks the median.
#'
#' @param data Data frame created by `quantiles_pbci`.
#' @param qseq Sequence of quantiles to plot - assumes median is in the middle of the sequence - default = deciles.
#'
#' @seealso \code{\link{quantiles_pbci}}
#'
#' @examples
#' set.seed(7)
#' # make fake skewed data
#' x <- rgamma(100, shape=3, scale=1)*10+250
#' # compute quantiles and their percentile bootstrap confidence intervals
#' out <- quantiles_pbci(x,q=seq(1,9)/10,nboot=2000,alpha=0.05)
#' # make decile plot
#' p <- plot_hd_ci(data=out,plotzero=TRUE,label.x="Onsets in ms",
#'                    hjust=-.05,vjust=.5,size_text=5,
#'                    colour_dec = "grey10",fill_dec = "grey90",
#'                    colour_line = "grey10", linetype_line = 1, size_line = 1) +
#'                    scale_y_continuous(limits=c(250, 350),breaks=seq(250,350,25))
# p
#'
#' @export
plot_hd_ci <- function(data = out,
                       qseq = seq(.1,.9,.1),
                        plotzero = TRUE,
                        label.x = "Differences",
                        hjust = -.05,
                        vjust = .2,
                        size_text = 6,
                        colour_q = "#009E73",
                        fill_q = "white",
                        size_q = 4,
                        shape_q = 21,
                        colour_line = "#009E73",
                        size_line = 1,
                        linetype_line = 2,
                        colour_zero = "black",
                        size_zero = .5,
                        linetype_zero = 1){
  md.loc <- floor(length(data$quantile)/2) + 1
  md <- data$est_q[md.loc] # median
  md.c <- as.character(round(md, digits=1)) # turn into characters
  lo.c <- as.character(round(data$ci.low[md.loc], digits=1)) # turn into characters
  up.c <- as.character(round(data$ci.up[md.loc], digits=1)) # turn into characters
  caption <- paste("Median = \n ",md.c," [",lo.c,", ",up.c,"]",sep="")
  if (all.equal(qseq,seq(.1,.9,.1))){
    label.y <- "Deciles"
  }  else {
    label.y <- "Quantiles"
  }
  p <- ggplot(data=out, aes(x=quantile*10, y=est_q)) +
    geom_abline(intercept = md, slope = 0,
                colour = colour_line,
                size = size_line,
                linetype = linetype_line)
  if (plotzero){
    p <- p + geom_abline(intercept = 0, slope = 0,
                         colour = colour_zero,
                         size = size_zero,
                         linetype = linetype_zero)
  }
  p <- p + geom_linerange(aes(ymin=ci.low, ymax=ci.up), colour=colour_line,size=1) +
    geom_point(colour = colour_q,
               size = size_q,
               shape = shape_q,
               fill = fill_q) +
    theme_bw() +
    labs(x=label.y) +
    labs(y=label.x) +
    theme(axis.text.x = element_text(size=14),
          axis.text.y = element_text(size=14),
          axis.title.x = element_text(size=16,face="bold"),
          axis.title.y = element_text(size=16,face="bold")) +
    scale_x_continuous(breaks=seq(1,9,1)) +
    annotate("text", x = 5, y = data$ci.up[5], label = caption[1],
             hjust = hjust, vjust = vjust, size = size_text) +
    coord_flip()
  p
}





# ----------------------------------------------------------------------------
#' Plot paired observations
#'
#' Scatterplot of paired observations with reference line of no effect.
#' Quartiles of each condition are superimposed. Quartiles are estimated using the
#' Harrell-Davis estimator.
#' @param df Data frame with paired observations in two columns.
#' @param formula A formula with format response variable ∼ predictor variable,
#'   where ~ (tilde) means "is modeled as a function of".
#' @param axis.steps Steps between x and y tick marks - default = 2.
#' @param min.x,min.y,max.x,max.y Specify axis limits - default square axes
#' @param colour_p Colour parameter of the scatterplot - default "black".
#' @param size_p Size parameter of the scatterplot - default = 5.
#' @param stroke_p Stroke parameter of the scatterplot - default = 1,
#' @param shape_p Shape parameter of the scatterplot - default = 21,
#' @param colour_p Colour parameter of the scatterplot - default = "black",
#' @param fill_p Fill parameter of the scatterplot - default = "#ffb347",
#' @param alpha_p Alpha parameter of the scatterplot - default = .5,
#' @param linetype_q Linetype of the segments marking the quartiles - default = "dashed",
#' @param size_q Size of the segments marking the quartiles - default = 1,
#' @param alpha_q Alpha of the segments marking the quartiles - default = .5,
#' @param colour_q Colour of the segments marking the quartiles - default = "black"
#' @examples
#' df <- tibble(cond1 = rnorm(50), cond2 = cond1 + rnorm(50))
#' plot_scat2d(df, formula = cond2 ~ cond1) # basic call
#' plot_scat2d(df, formula = cond2 ~ cond1, size_q=3) # specify size of quartile segments
#' plot_scat2d(df, formula = cond2 ~ cond1, size_q=c(1,2,1)) # use thicker line for median
#' plot_scat2d(df, formula = cond2 ~ cond1, linetype_q = "longdash") # specify linetype - default = dashed
#' @seealso \code{\link{hd}}
#' @export
plot_scat2d <- function(df = df,
                        formula = cond2 ~ cond1,
                        min.x = NA,
                        min.y = NA,
                        max.x = NA,
                        max.y = NA,
                        axis.steps = 2,
                        size_p = 5,
                        stroke_p = 1,
                        shape_p = 21,
                        colour_p = "black",
                        fill_p = "#ffb347",
                        alpha_p = .5,
                        linetype_q = "dashed",
                        size_q = 1,
                        alpha_q = .5,
                        colour_q = "black"){
  # subset formula
  subf <- subset_formula_wide(df, formula)
  xplot = subf$x_col_name
  yplot = subf$y_col_name
  # make data.frames for plotting quartile segments
  hd1.25 <- hd(df[[xplot]],.25)
  hd1.5 <- hd(df[[xplot]],.5)
  hd1.75 <- hd(df[[xplot]],.75)
  hd2.25 <- hd(df[[yplot]],.25)
  hd2.5 <- hd(df[[yplot]],.5)
  hd2.75 <- hd(df[[yplot]],.75)
  df.25 <- data.frame(hd1=hd1.25,hd2=hd2.25)
  df.5 <- data.frame(hd1=hd1.5,hd2=hd2.5)
  df.75 <- data.frame(hd1=hd1.75,hd2=hd2.75)

  # quartile plot parameters
  if(length(linetype_q)==1){
    linetype_q = rep(linetype_q,3)
  }
  if(length(size_q)==1){
    size_q = rep(size_q,3)
  }
  if(length(alpha_q)==1){
    alpha_q = rep(alpha_q,3)
  }
  if(length(colour_q)==1){
    colour_q = rep(colour_q,3)
  }

  # plot limits
  if (is.na(min.x)){
    min.x <- min(df[[xplot]],df[[yplot]])
  }
  if (is.na(max.x)){
    max.x <- max(df[[xplot]],df[[yplot]])
  }
  if (is.na(min.y)){
    min.y <- min(df[[xplot]],df[[yplot]])
  }
  if (is.na(max.y)){
    max.y <- max(df[[xplot]],df[[yplot]])
  }

  # scatterplot of paired observations -----------------
  p <- ggplot(df, aes_string(x = xplot, y = yplot)) +
    # reference line
    geom_abline(intercept = 0) +
    # quartiles
    scale_x_continuous(breaks=seq(floor(min.x),ceiling(max.x),axis.steps)) +
    scale_y_continuous(breaks=seq(floor(min.y),ceiling(max.y),axis.steps)) +
    coord_cartesian(xlim = c(floor(min.x), ceiling(max.x)),
                    ylim = c(floor(min.y), ceiling(max.y))) +
    geom_segment(aes(x=hd1,y=min.y-abs(min.y),xend=hd1,yend=hd2),data=df.25,linetype=linetype_q[1],size=size_q[1],alpha=alpha_q[1],colour=colour_q[1]) +
    geom_segment(aes(x=hd1,y=min.y-abs(min.y),xend=hd1,yend=hd2),data=df.5,linetype=linetype_q[2],size=size_q[2],alpha=alpha_q[2],colour=colour_q[2]) +
    geom_segment(aes(x=hd1,y=min.y-abs(min.y),xend=hd1,yend=hd2),data=df.75,linetype=linetype_q[3],size=size_q[3],alpha=alpha_q[3],colour=colour_q[3]) +
    geom_segment(aes(x=min.x-abs(min.x),y=hd2,xend=hd1,yend=hd2),data=df.25,linetype=linetype_q[1],size=size_q[1],alpha=alpha_q[1],colour=colour_q[1]) +
    geom_segment(aes(x=min.x-abs(min.x),y=hd2,xend=hd1,yend=hd2),data=df.5,linetype=linetype_q[2],size=size_q[2],alpha=alpha_q[2],colour=colour_q[2]) +
    geom_segment(aes(x=min.x-abs(min.x),y=hd2,xend=hd1,yend=hd2),data=df.75,linetype=linetype_q[3],size=size_q[3],alpha=alpha_q[3],colour=colour_q[3]) +
    # scatterplot
    geom_point(size=size_p,
               stroke=stroke_p,
               shape=shape_p,
               colour=colour_p,
               fill=fill_p,
               alpha=alpha_p) +
    theme_bw() +
    theme(axis.text.x = element_text(size=14),
          axis.text.y = element_text(size=14),
          axis.title.x = element_text(size=16,face="bold"),
          axis.title.y = element_text(size=16,face="bold"),
          legend.title = element_blank(),
          plot.title = element_text(size=20)) +
    labs(title="Paired observations")
  p
}
# ----------------------------------------------------------------------------
